File: apps/ai-service/app/services/providers/anthropic_provider.py
"""
Anthropic Provider Implementation
Supports: Claude 3.5 Sonnet, Claude 3 Opus, Claude 3 Haiku
"""

import asyncio
import logging
from typing import List, Dict, Any, Optional, AsyncIterator
import anthropic
from anthropic import Anthropic, AsyncAnthropic
import os
import time

from .base_provider import (
    AIProvider,
    ProviderType,
    ProviderConfig,
    ModelCapability,
    ModelInfo,
    TranscriptionRequest,
    TranscriptionResponse,
    ChatRequest,
    ChatResponse,
    ChatMessage,
    EmbeddingRequest,
    EmbeddingResponse,
    VisionRequest,
    VisionResponse,
)

logger = logging.getLogger(__name__)


class AnthropicProvider(AIProvider):
    """Anthropic (Claude) API provider implementation"""

    def __init__(self, config: ProviderConfig):
        super().__init__(config)

        # Initialize Anthropic clients
        api_key = config.api_key or os.getenv("ANTHROPIC_API_KEY")
        if not api_key:
            raise ValueError("Anthropic API key not provided")

        self.client = Anthropic(
            api_key=api_key,
            base_url=config.api_base,
            timeout=config.timeout,
            max_retries=config.max_retries,
        )

        self.async_client = AsyncAnthropic(
            api_key=api_key,
            base_url=config.api_base,
            timeout=config.timeout,
            max_retries=config.max_retries,
        )

        # Model registry
        self.models = {
            "claude-3-5-sonnet": ModelInfo(
                model_id="claude-3-5-sonnet-20241022",
                name="Claude 3.5 Sonnet",
                provider=ProviderType.ANTHROPIC,
                capabilities=[ModelCapability.CHAT, ModelCapability.TEXT_GENERATION, ModelCapability.VISION],
                context_window=200000,
                max_output_tokens=8192,
                cost_per_1k_input=0.003,
                cost_per_1k_output=0.015,
                supports_streaming=True,
                supports_function_calling=True,
                description="Most intelligent model, best for complex tasks",
            ),
            "claude-3-opus": ModelInfo(
                model_id="claude-3-opus-20240229",
                name="Claude 3 Opus",
                provider=ProviderType.ANTHROPIC,
                capabilities=[ModelCapability.CHAT, ModelCapability.TEXT_GENERATION, ModelCapability.VISION],
                context_window=200000,
                max_output_tokens=4096,
                cost_per_1k_input=0.015,
                cost_per_1k_output=0.075,
                supports_streaming=True,
                supports_function_calling=True,
                description="Top-level performance on highly complex tasks",
            ),
            "claude-3-sonnet": ModelInfo(
                model_id="claude-3-sonnet-20240229",
                name="Claude 3 Sonnet",
                provider=ProviderType.ANTHROPIC,
                capabilities=[ModelCapability.CHAT, ModelCapability.TEXT_GENERATION, ModelCapability.VISION],
                context_window=200000,
                max_output_tokens=4096,
                cost_per_1k_input=0.003,
                cost_per_1k_output=0.015,
                supports_streaming=True,
                supports_function_calling=True,
                description="Balance of intelligence and speed",
            ),
            "claude-3-haiku": ModelInfo(
                model_id="claude-3-haiku-20240307",
                name="Claude 3 Haiku",
                provider=ProviderType.ANTHROPIC,
                capabilities=[ModelCapability.CHAT, ModelCapability.TEXT_GENERATION, ModelCapability.VISION],
                context_window=200000,
                max_output_tokens=4096,
                cost_per_1k_input=0.00025,
                cost_per_1k_output=0.00125,
                supports_streaming=True,
                supports_function_calling=True,
                description="Fastest and most compact model",
            ),
        }

        logger.info(f"✅ Anthropic Provider initialized ({len(self.models)} models)")

    async def transcribe(self, request: TranscriptionRequest) -> TranscriptionResponse:
        """Anthropic doesn't support transcription"""
        raise NotImplementedError("Anthropic doesn't provide transcription. Use OpenAI or Local provider.")

    async def chat_completion(self, request: ChatRequest) -> ChatResponse:
        """Generate chat completion with Claude"""
        try:
            model = request.model or "claude-3-5-sonnet-20241022"

            # Convert messages (Claude format different from OpenAI)
            system_messages = [msg.content for msg in request.messages if msg.role == "system"]
            system = "\n\n".join(system_messages) if system_messages else None

            messages = [
                {"role": msg.role, "content": msg.content}
                for msg in request.messages
                if msg.role in ["user", "assistant"]
            ]

            response = await self.async_client.messages.create(
                model=model,
                system=system,
                messages=messages,
                temperature=request.temperature,
                max_tokens=request.max_tokens or 4096,
                top_p=request.top_p,
                stop_sequences=request.stop,
                stream=False,
            )

            return ChatResponse(
                content=response.content[0].text,
                model=response.model,
                provider="anthropic",
                finish_reason=response.stop_reason,
                usage={
                    "prompt_tokens": response.usage.input_tokens,
                    "completion_tokens": response.usage.output_tokens,
                    "total_tokens": response.usage.input_tokens + response.usage.output_tokens,
                },
            )

        except Exception as e:
            logger.error(f"Anthropic chat completion failed: {e}")
            raise

    async def chat_completion_stream(self, request: ChatRequest) -> AsyncIterator[str]:
        """Generate streaming chat completion"""
        try:
            model = request.model or "claude-3-5-sonnet-20241022"

            system_messages = [msg.content for msg in request.messages if msg.role == "system"]
            system = "\n\n".join(system_messages) if system_messages else None

            messages = [
                {"role": msg.role, "content": msg.content}
                for msg in request.messages
                if msg.role in ["user", "assistant"]
            ]

            async with self.async_client.messages.stream(
                model=model,
                system=system,
                messages=messages,
                temperature=request.temperature,
                max_tokens=request.max_tokens or 4096,
                top_p=request.top_p,
            ) as stream:
                async for text in stream.text_stream:
                    yield text

        except Exception as e:
            logger.error(f"Anthropic streaming failed: {e}")
            raise

    async def generate_embedding(self, request: EmbeddingRequest) -> EmbeddingResponse:
        """Anthropic doesn't support embeddings"""
        raise NotImplementedError("Anthropic doesn't provide embeddings. Use OpenAI or Local provider.")

    async def vision_completion(self, request: VisionRequest) -> VisionResponse:
        """Process vision request with Claude"""
        try:
            model = request.model or "claude-3-5-sonnet-20241022"

            # Read image and encode to base64
            import base64
            with open(request.image_path, "rb") as f:
                image_data = base64.b64encode(f.read()).decode("utf-8")

            # Detect media type
            import mimetypes
            media_type = mimetypes.guess_type(request.image_path)[0] or "image/jpeg"

            response = await self.async_client.messages.create(
                model=model,
                max_tokens=request.max_tokens or 1024,
                messages=[
                    {
                        "role": "user",
                        "content": [
                            {
                                "type": "image",
                                "source": {
                                    "type": "base64",
                                    "media_type": media_type,
                                    "data": image_data,
                                },
                            },
                            {
                                "type": "text",
                                "text": request.prompt,
                            },
                        ],
                    }
                ],
            )

            return VisionResponse(
                content=response.content[0].text,
                model=response.model,
                provider="anthropic",
            )

        except Exception as e:
            logger.error(f"Anthropic vision failed: {e}")
            raise

    def list_models(self) -> List[ModelInfo]:
        """List available models"""
        return list(self.models.values())

    async def health_check(self) -> bool:
        """Check Anthropic API health"""
        try:
            # Simple test request
            await self.async_client.messages.create(
                model="claude-3-haiku-20240307",
                max_tokens=10,
                messages=[{"role": "user", "content": "Hi"}],
            )
            return True
        except Exception as e:
            logger.warning(f"Anthropic health check failed: {e}")
            return False

    def get_capabilities(self) -> List[ModelCapability]:
        """Get provider capabilities"""
        return [
            ModelCapability.CHAT,
            ModelCapability.TEXT_GENERATION,
            ModelCapability.VISION,
        ]

    async def estimate_cost(
        self,
        request: Union[ChatRequest, TranscriptionRequest],
        model: Optional[str] = None
    ) -> Dict[str, float]:
        """Estimate cost for request"""
        if isinstance(request, ChatRequest):
            model_id = model or request.model or "claude-3-5-sonnet-20241022"
            model_info = self.models.get(model_id.split("/")[-1])

            if not model_info:
                return {"input_cost": 0.0, "output_cost": 0.0, "total_cost": 0.0}

            # Token estimation
            total_chars = sum(len(msg.content) for msg in request.messages)
            estimated_input_tokens = total_chars // 4
            estimated_output_tokens = request.max_tokens or 1000

            input_cost = (estimated_input_tokens / 1000) * model_info.cost_per_1k_input
            output_cost = (estimated_output_tokens / 1000) * model_info.cost_per_1k_output
            total_cost = input_cost + output_cost

            return {
                "input_cost": round(input_cost, 6),
                "output_cost": round(output_cost, 6),
                "total_cost": round(total_cost, 6),
                "estimated_input_tokens": estimated_input_tokens,
                "estimated_output_tokens": estimated_output_tokens,
            }

        return {"input_cost": 0.0, "output_cost": 0.0, "total_cost": 0.0}

File: apps/api/src/services/ai/providers/AnthropicProvider.ts
/**
 * Anthropic Provider
 *
 * Real Anthropic SDK integration supporting:
 * - Claude-3 Opus, Sonnet, Haiku
 * - Streaming support
 * - Message batching
 * - Vision capabilities
 */

import Anthropic from '@anthropic-ai/sdk';
import * as winston from 'winston';
import {
  IAIProvider,
  AIProviderType,
  ProviderCapabilities,
  ProviderConfig,
  ChatCompletionOptions,
  ChatCompletionResponse,
  EmbeddingOptions,
  EmbeddingResponse,
} from '../AIProviderFactory';

const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || 'info',
  defaultMeta: { service: 'anthropic-provider' },
  transports: [new winston.transports.Console()],
});

/**
 * Anthropic Provider Implementation
 */
export class AnthropicProvider implements IAIProvider {
  readonly type: AIProviderType = 'anthropic';
  readonly capabilities: ProviderCapabilities = {
    chat: true,
    embeddings: false, // Anthropic doesn't provide embeddings
    transcription: false, // Anthropic doesn't provide transcription
    vision: true,
    streaming: true,
    functions: true, // Claude supports tool use
  };

  private client: Anthropic;
  private config: Required<ProviderConfig>;

  constructor(config?: ProviderConfig) {
    // Load configuration with defaults
    this.config = {
      apiKey: config?.apiKey || process.env.ANTHROPIC_API_KEY || '',
      baseURL: config?.baseURL || process.env.ANTHROPIC_BASE_URL || 'https://api.anthropic.com',
      model: config?.model || process.env.ANTHROPIC_MODEL || 'claude-3-5-sonnet-20241022',
      embeddingModel: config?.embeddingModel || '', // Not supported
      timeout: config?.timeout || 60000,
      maxRetries: config?.maxRetries || 3,
    };

    // Validate API key
    if (!this.config.apiKey) {
      throw new Error('Anthropic API key is required. Set ANTHROPIC_API_KEY environment variable.');
    }

    // Initialize Anthropic client
    this.client = new Anthropic({
      apiKey: this.config.apiKey,
      baseURL: this.config.baseURL,
      timeout: this.config.timeout,
      maxRetries: this.config.maxRetries,
    });

    logger.info('Anthropic provider initialized', {
      baseURL: this.config.baseURL,
      model: this.config.model,
    });
  }

  /**
   * Get available Anthropic models
   */
  async getAvailableModels(): Promise<string[]> {
    // Anthropic doesn't have a models list endpoint, return known models
    const models = [
      'claude-3-5-sonnet-20241022',
      'claude-3-5-haiku-20241022',
      'claude-3-opus-20240229',
      'claude-3-sonnet-20240229',
      'claude-3-haiku-20240307',
    ];

    logger.debug(`Available Anthropic models: ${models.length}`);
    return models;
  }

  /**
   * Chat completion with Claude models
   */
  async chatCompletion(options: ChatCompletionOptions): Promise<ChatCompletionResponse> {
    const {
      messages,
      temperature = 0.7,
      maxTokens = 1000,
      stream = false,
      topP,
      stop,
    } = options;

    try {
      const startTime = Date.now();

      // Extract system message if present
      const systemMessage = messages.find(m => m.role === 'system');
      const conversationMessages = messages.filter(m => m.role !== 'system');

      // Convert messages to Anthropic format
      const anthropicMessages: Anthropic.MessageParam[] = conversationMessages.map(msg => ({
        role: msg.role === 'assistant' ? 'assistant' : 'user',
        content: msg.content,
      }));

      // Create message
      const response = await this.client.messages.create({
        model: this.config.model,
        max_tokens: maxTokens,
        messages: anthropicMessages,
        system: systemMessage?.content,
        temperature,
        top_p: topP,
        stop_sequences: stop,
        stream: false,
      });

      const duration = Date.now() - startTime;

      // Extract content from response
      const contentBlock = response.content[0];
      if (!contentBlock || contentBlock.type !== 'text') {
        throw new Error('No text content in Anthropic response');
      }

      const result: ChatCompletionResponse = {
        content: contentBlock.text,
        model: response.model,
        usage: response.usage
          ? {
              promptTokens: response.usage.input_tokens,
              completionTokens: response.usage.output_tokens,
              totalTokens: response.usage.input_tokens + response.usage.output_tokens,
            }
          : undefined,
        finishReason: response.stop_reason || undefined,
      };

      logger.info('Anthropic chat completion successful', {
        model: response.model,
        duration,
        tokens: result.usage?.totalTokens,
        finishReason: result.finishReason,
      });

      return result;
    } catch (error) {
      logger.error('Anthropic chat completion failed', {
        model: this.config.model,
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      throw error;
    }
  }

  /**
   * Generate embeddings (not supported by Anthropic)
   */
  async generateEmbeddings(options: EmbeddingOptions): Promise<EmbeddingResponse> {
    throw new Error('Anthropic does not support embeddings. Use OpenAI or another provider.');
  }

  /**
   * Test connection to Anthropic
   */
  async testConnection(): Promise<boolean> {
    try {
      // Try a minimal message to test the connection
      await this.client.messages.create({
        model: this.config.model,
        max_tokens: 10,
        messages: [{ role: 'user', content: 'Hi' }],
      });

      logger.debug('Anthropic connection test successful');
      return true;
    } catch (error) {
      logger.warn('Anthropic connection test failed', {
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      return false;
    }
  }

  /**
   * Get current configuration
   */
  getConfig(): ProviderConfig {
    return {
      apiKey: '***redacted***',
      baseURL: this.config.baseURL,
      model: this.config.model,
      timeout: this.config.timeout,
      maxRetries: this.config.maxRetries,
    };
  }

  /**
   * Chat completion with streaming (for future use)
   */
  async *chatCompletionStream(
    options: ChatCompletionOptions
  ): AsyncGenerator<string, void, unknown> {
    const {
      messages,
      temperature = 0.7,
      maxTokens = 1000,
      topP,
      stop,
    } = options;

    try {
      // Extract system message if present
      const systemMessage = messages.find(m => m.role === 'system');
      const conversationMessages = messages.filter(m => m.role !== 'system');

      // Convert messages to Anthropic format
      const anthropicMessages: Anthropic.MessageParam[] = conversationMessages.map(msg => ({
        role: msg.role === 'assistant' ? 'assistant' : 'user',
        content: msg.content,
      }));

      // Create streaming message
      const stream = await this.client.messages.create({
        model: this.config.model,
        max_tokens: maxTokens,
        messages: anthropicMessages,
        system: systemMessage?.content,
        temperature,
        top_p: topP,
        stop_sequences: stop,
        stream: true,
      });

      for await (const event of stream) {
        if (event.type === 'content_block_delta' && event.delta.type === 'text_delta') {
          yield event.delta.text;
        }
      }
    } catch (error) {
      logger.error('Anthropic streaming completion failed', {
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      throw error;
    }
  }

  /**
   * Create message with tools (function calling)
   */
  async chatCompletionWithTools(
    options: ChatCompletionOptions & {
      tools?: Anthropic.Tool[];
    }
  ): Promise<ChatCompletionResponse & { toolCalls?: any[] }> {
    const {
      messages,
      temperature = 0.7,
      maxTokens = 1000,
      topP,
      stop,
      tools,
    } = options;

    try {
      const startTime = Date.now();

      // Extract system message if present
      const systemMessage = messages.find(m => m.role === 'system');
      const conversationMessages = messages.filter(m => m.role !== 'system');

      // Convert messages to Anthropic format
      const anthropicMessages: Anthropic.MessageParam[] = conversationMessages.map(msg => ({
        role: msg.role === 'assistant' ? 'assistant' : 'user',
        content: msg.content,
      }));

      // Create message with tools
      const response = await this.client.messages.create({
        model: this.config.model,
        max_tokens: maxTokens,
        messages: anthropicMessages,
        system: systemMessage?.content,
        temperature,
        top_p: topP,
        stop_sequences: stop,
        tools,
      });

      const duration = Date.now() - startTime;

      // Extract content and tool calls
      let textContent = '';
      const toolCalls: any[] = [];

      for (const block of response.content) {
        if (block.type === 'text') {
          textContent += block.text;
        } else if (block.type === 'tool_use') {
          toolCalls.push({
            id: block.id,
            name: block.name,
            input: block.input,
          });
        }
      }

      const result = {
        content: textContent,
        model: response.model,
        usage: response.usage
          ? {
              promptTokens: response.usage.input_tokens,
              completionTokens: response.usage.output_tokens,
              totalTokens: response.usage.input_tokens + response.usage.output_tokens,
            }
          : undefined,
        finishReason: response.stop_reason || undefined,
        toolCalls: toolCalls.length > 0 ? toolCalls : undefined,
      };

      logger.info('Anthropic chat completion with tools successful', {
        model: response.model,
        duration,
        tokens: result.usage?.totalTokens,
        toolCalls: toolCalls.length,
      });

      return result;
    } catch (error) {
      logger.error('Anthropic chat completion with tools failed', {
        model: this.config.model,
        error: error instanceof Error ? error.message : 'Unknown error',
      });
      throw error;
    }
  }
}
